cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java

```diff
@@ -17,51 +17,63 @@
 package co.cask.cdap.data2.nosql;
 
 import co.cask.cdap.api.dataset.DatasetAdmin;
-import co.cask.cdap.api.dataset.DatasetManagementException;
+import co.cask.cdap.api.dataset.DatasetContext;
+import co.cask.cdap.api.dataset.DatasetDefinition;
 import co.cask.cdap.api.dataset.DatasetProperties;
-import co.cask.cdap.api.dataset.table.Table;
-import co.cask.cdap.data2.dataset2.DatasetFramework;
-import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.api.dataset.module.DatasetDefinitionRegistry;
+import co.cask.cdap.api.dataset.module.DatasetModule;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.dataset2.InMemoryDatasetDefinitionRegistry;
 import co.cask.cdap.proto.id.NamespaceId;
 import co.cask.cdap.spi.data.StructuredTableAdmin;
 import co.cask.cdap.spi.data.table.StructuredTableId;
 import co.cask.cdap.spi.data.table.StructuredTableSpecification;
 import co.cask.cdap.spi.data.table.StructuredTableSpecificationRegistry;
+import com.google.inject.Injector;
+import com.google.inject.Key;
+import com.google.inject.name.Names;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.util.Collections;
 
 /**
  * The Nosql admin will use the existing dataset framework to create and drop tables.
  */
 public final class NoSqlStructuredTableAdmin implements StructuredTableAdmin {
-  static final String ENTITY_TABLE_NAME = "entity.store";
+
   private static final Logger LOG = LoggerFactory.getLogger(NoSqlStructuredTableAdmin.class);
+  private static final DatasetContext SYSTEM_CONTEXT = DatasetContext.from(NamespaceId.SYSTEM.getNamespace());
+
+  static final String ENTITY_TABLE_NAME = "entity.store";
 
-  private final DatasetFramework datasetFramework;
+  private final DatasetDefinition tableDefinition;
+  private final DatasetSpecification entityTableSpec;
 
-  public NoSqlStructuredTableAdmin(DatasetFramework datasetFramework) {
-    this.datasetFramework = datasetFramework;
+  public NoSqlStructuredTableAdmin(Injector injector) {
```

Comment:
It is better to annotate this constructor with `@Inject` and take only the the necessary dependency (`@Named(Constants.Dataset.TABLE_MODULE) DatasetModule tableModule)` in this case). Then instead of "newing" this class directly, you can get an instance of it through the injector

```java
NoSqlStructuredTableAdmin admin = injector.getInstance(NoSqlStructuredTableAdmin.class);
```

Reply:
I tried that. The issue with that is that the module, after registring itself, provides a DatasetDefinition for the Table. On that, I have to call the injector again to inject members. I could still do that in the way you described, but in any case I need the injector itself in the table admin.

Reply:
Let me push a commit that does t this way and see whether that looks better.

Reply:
I see. In that case, you can do a `@Inject` that takes an `Injector`. Better than creating an instance directly.

Reply:
Especially works better for private binding.